text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,426 @@
+```
+
+Then the `Cargo.toml` can use those path bases and avoid relative paths:
+
+```toml
+[dependencies]
+foo = { path = "other_layer/foo", base = "sources" }
+bar_with_generated = { path = "third_layer/bar_with_generated", base = "intermediates" }
+baz = { path = "this_layer/baz", base = "sources" }
+```
+
+Which resolves the issues we previously had:
+
+* The current project can be moved without modifying the `Cargo.toml` at all.
+* `bar_with_generated` works for all targets (assuming the `config.toml` is
+
+generated).
+* The `intermediates` directory can be placed anywhere.
+* Moving `foo` or `baz` only requires searching for the canonical form relative
+to the path base.
+
+## Other uses
+
+The ability to use path bases for `path` dependencies is convenient for
+developers who are using a large number of `path` dependencies within the same
+root directory. Instead of repeating the same path fragment many times in their
+`Cargo.toml`, they can instead specify it once in a `config.toml` as a path
+base, then use that path base in each of their `path` dependencies.
+
+Cargo can also provide built-in base paths, for example `workspace` to point to
+the root directory of the workspace. This allows workspace members to reference
+each other without first needing to `../` their way back to the workspace root.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+If you often use multiple path dependencies that have a common parent directory,
+or if you want to avoid putting long paths in your `Cargo.toml`, you can
+define path _base directories_ in your
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html).
+Your path dependencies can then be specified relative to those base
+directories.
+
+For example, say you have a number of projects checked out in
+`/home/user/dev/rust/libraries/`. Rather than use that path in your
+`Cargo.toml` files, you can define it as a "base" path in
+`~/.cargo/config.toml`:
+
+```toml
+[path-bases]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+Now, you can specify a path dependency on a library `foo` in that
+directory in your `Cargo.toml` using
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Like with other path dependencies, keep in mind that both the base _and_
+the path must exist on any other host where you want to use the same
+`Cargo.toml` to build your project.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Specifying Dependencies
+
+### Path Bases
+
+A `path` dependency may optionally specify a base by setting the `base` key to
+the name of a path base from the `[path-bases]` table in either the
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html#path-bases)
+or one of the [built-in path bases](#built-in-path-bases). The value of that
+path base is prepended to the `path` value to produce the actual location where

```

Comment:
```suggestion
path base is prepended to the `path` value (along with a path separator if necessary) to produce the actual location where

```

(This ensures a base of `xyz` and a path of `abc` produces `xyz/abc` rather than `xyzabc`.)